error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/ClassCastTest.java

```diff
@@ -0,0 +1,62 @@
+package tech.picnic.errorprone.bugpatterns;
+
+import com.google.errorprone.BugCheckerRefactoringTestHelper;
+import com.google.errorprone.BugCheckerRefactoringTestHelper.TestMode;
+import com.google.errorprone.CompilationTestHelper;
+import org.junit.jupiter.api.Test;
+
+final class ClassCastTest {
+  @Test
+  void identification() {
+    CompilationTestHelper.newInstance(ClassCast.class, getClass())
+        .addSourceLines(
+            "A.java",
+            "import com.google.common.collect.ImmutableList;",
+            "import com.google.common.collect.ImmutableSet;",
+            "import java.util.Optional;",
+            "import java.util.Set;",
+            "",
+            "class A {",
+            "  void m() {",
+            "    Number foo = 0;",
+            "    Number bar = 1;",
+            "",
+            "    // BUG: Diagnostic contains:",
+            "    Optional.of(foo).map(i -> (Integer) i);",
+            "",
+            "    Optional.of(foo).map(i -> 2);",
+            "    Optional.of(foo).map(i -> (Integer) 2);",
+            "    Optional.of(foo).map(i -> bar);",
+            "    Optional.of(foo).map(i -> (Integer) bar);",
+            "",
+            "    ImmutableList.of(Set.of(foo)).stream().map(l -> (ImmutableSet<Number>) l);",
+            "    ImmutableList.of(Set.of(foo)).stream().map(l -> (ImmutableSet<?>) l);",
```

Comment:
Would have been nice to stick a bit closer to the test cases in `IsInstanceLambdaUsageTest`. W.r.t. the `ImmutableSet<Number>` case: let's more explicitly generate a stream of raw type elements.